vm/src/compile.rs

```diff
@@ -527,31 +530,6 @@ impl Compiler {
         self.emit(Instruction::BinaryOperation { op: i });
     }
 
-    fn compile_test(&mut self, expression: &ast::Expression, not_label: Label) {
```

Comment:
Why was this function removed? So the idea was that compile_test will jump to two labels, and execute that code. Now for each `or` and `and` operation first a boolean is put onto the stack. This maybe not required.

Reply:
Because tests don't know the context they're compiled in, and shouldn't. The jumps belong to context surrounding the test. (Many tests are compiled in contexts with no final jumps) I think this implementation is fairly simple and the passing labels around is going to get really complex.

Reply:
Okay, I do not understand this. If you like, can you give an example where this is an issue? This implementation is indeed fairly simple and works, so its fine.

Reply:
I think I've figured out the difference. In the following expression:

```False and False and False and False```

My version has to jump several several times, your version can make a single jump when it evaluates the first false.

CPython does indeed produce the version with the single jump:
```
  1           0 LOAD_CONST               1 (False)
              3 JUMP_IF_FALSE_OR_POP    21
              6 LOAD_CONST               1 (False)
              9 JUMP_IF_FALSE_OR_POP    21
             12 LOAD_CONST               1 (False)
             15 JUMP_IF_FALSE_OR_POP    21
             18 LOAD_CONST               1 (False)
        >>   21 RETURN_VALUE
```
The fun thing is: python is a weird language where we can write programs that depend on this optimisation:

```
class LoggedFalse():
    def __init__(self):
        self.booled = 0

    def __bool__(self):
        self.booled += 1
        print("It's False!")
        return False

y = (LoggedFalse() and False and True and False)

print(y)

assert y.booled == 1
```

$ python3 tests/snippets/shortcuts.py 
It's False!
<__main__.LoggedFalse object at 0x7fce9820aac8>
$ cargo run tests/snippets/shortcuts.py
It's False! 
It's False! 
It's False! 
<instance> 
thread 'main' panicked at 'Exception: RefCell { value: [PyObj instance] }', src/main.rs:61:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Now that I understand why passing labels is important, I'd like to have a go at implementing that version.

Reply:
Thanks for the detailed explanation! I first implement the short-circuit jumps to let this code pass:

```python
x = False and undefined_symbol1
```

Also, this code would result in an immediate jump to the else clause, not having to put an intermediate value on the value stack, and then checking this value with `JumpIfFalse`
```python
if False and undefined_symbol2:
    print(1)
else:
    print(2)
```

Reply:
Your first example does pass in my (the current) version. The version with the if does indeed evaluate the False twice.

When I realised I could write a program that would distinguish this optimisation, I assumed it would be an invalid optimisation, not a mandatory one.

It's weird that these two programs don't necessarily behave the same:

```
c = a and b
assert c
```

```
assert a and b
```

Thanks for getting me to think more about this.